File: ta_lib/strategies/signal/src/ma/vwap_cross.rs
use base::prelude::*;
use core::prelude::*;
use volume::vwap;

pub struct VwapCrossSignal {
    source_type: SourceType,
    period: usize,
}

impl VwapCrossSignal {
    pub fn new(source_type: SourceType, period: f32) -> Self {
        Self {
            source_type,
            period: period as usize,
        }
    }
}

impl Signal for VwapCrossSignal {
    fn lookback(&self) -> usize {
        self.period
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let vwap = vwap(&data.source(self.source_type), data.volume());

        (
            data.close().cross_over(&vwap),
            data.close().cross_under(&vwap),
        )
    }
}

File: factor/signal/ma/vwap_cross.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter
from core.models.source import SourceType
from factor.signal.base import Signal, SignalType


@dataclass(frozen=True)
class VwapCrossSignal(Signal):
    type: SignalType = SignalType.VwapCross
    source_type: Parameter = StaticParameter(SourceType.HLC3)
    period: Parameter = StaticParameter(100.0)
